### PR TITLE
Integration of Nuget and Vsix into CI pipeline

### PR DESCRIPTION
**Pull request by middlemose**
Adds Nuget package capability to the CI pipeline.

Adds installation of WDK vsix directly to the CI pipeline.

Updates environment variables for ease of incrementing component versions.

### Pipeline Time Metrics
**BuildAll** (push and pull_request)
Without Vsix Install: ~08m 59s
With Vsix Install:    ~38m 59s (optimized ~8m)

**BuildChange** (pull_request)
Without Vsix Install: ~03m 17s
With Vsix Install:    ~23m 20s (optimized "_pending_")

**CodeQL** (push and pull_request)
Without Vsix Install: ~1h 1m 43s
With Vsix Install:    ~1h 34m 31s (optimized ~1h 16m 2s)

**Average WDK Vsix Install Time**
~21m (optimized ~15s)

**Average WDK Nuget Restore**
~1m 30s

image

image

### CODE REVIEW

File: .github/workflows/ci-pr.yml [hunk 1]
```diff
@@ -63,8 +66,7 @@ jobs:
           path: _logs
 
       - name: Join and generate global reports
-        run: |
```
Comment (JakobL-MSFT) [inline .github/workflows/ci-pr.yml:66]: Just for this specific comment.  What does the removal of all the "|" lines do?  I assume this is just cosmetic, but good time for me to learn.

**Comment by JakobL-MSFT** [comment]
Do you need to update your performance numbers now that you have $optimize?

**Review by JakobL-MSFT — CHANGES_REQUESTED** [review]

File: .github/workflows/ci-pr.yml (hunk 1 above)
Comment (middlemose) [reply .github/workflows/ci-pr.yml:66]: Just cosmetic as it was not a 'multi-line' command.
https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsrun:~:text=A%20multi%2Dline%20command%3A

File: .github/workflows/ci-pr.yml [hunk 2]
```diff
@@ -22,8 +22,11 @@ jobs:
         with:
           submodules: 'recursive'
 
-      - name: Add MSBuild to PATH
```
[the remark this answers is not in the document]
Comment (middlemose) [reply .github/workflows/ci-pr.yml:25]: The motivation was not to use an indirect way of building, but rather match the developer approach which is to use the Developer Command Prompt / Powershell (if not in the IDE). In my opinion this is a more guaranteed way of having all the proper VS and msbuild environment variables set. Plus, I had an early on issue when developing the install vsix without using developer environment.

File: .github/scripts/Env-Vars.ps1 [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# source this file to import these environment variables into a session
+$env:SAMPLES_VSIX_VERSION = "10.0.26100.0"
+$env:SAMPLES_VSIX_URI = "https://marketplace.visualstudio.com/_apis/public/gallery/publishers/DriverDeveloperKits-WDK/vsextensions/WDKVsix/10.0.26100.0/vspackage?targetPlatform=5e3e564c-03bb-4499-8ae5-b2b35e9a86dc"
+$env:SAMPLES_BUILD_NUMBER = "26100"
+$env:SAMPLES_NUGET_VERSION = "10.0.26100.1"
```
[the remark this answers is not in the document]
Comment (middlemose) [reply .github/scripts/Env-Vars.ps1:5]: You are correct. I will remove it. 

It was used when I approached generating the config and prop files.

File: general/echo/kmdf/driver/AutoSync/echo.inx
[the remark this answers is not in the document]
Comment (middlemose) [reply general/echo/kmdf/driver/AutoSync/echo.inx:1]: To trigger building only single sample and not all.

File: Build-SampleSet.ps1 [hunk 4]
```diff
@@ -9,6 +9,17 @@ param(
...
+
+# launch developer powershell (if necessary)
+if (-not $env:VSCMD_VER) {
+    Import-Module (Resolve-Path "$env:ProgramFiles\Microsoft Visual Studio\2022\*\Common7\Tools\Microsoft.VisualStudio.DevShell.dll")
+    Enter-VsDevShell -VsInstallPath (Resolve-Path "$env:ProgramFiles\Microsoft Visual Studio\2022\*")
+    cd $root
+}
+
+# source environment variables
+. .\.github\scripts\Env-Vars.ps1
```
[the remark this answers is not in the document]
Comment (middlemose) [reply Build-SampleSet.ps1:21]: I debated where to put this file. I can move it to root.

File: .github/workflows/ci-pr.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (JakobL-MSFT) [reply .github/workflows/ci-pr.yml:25]: Love it! Thank you.

File: general/echo/kmdf/driver/AutoSync/echo.inx
[the remark this answers is not in the document]
Comment (JakobL-MSFT) [reply general/echo/kmdf/driver/AutoSync/echo.inx:1]: Great, but that is just for testing purposes.  Once testing is complete please remove this change?

File: input/kbfiltr/sys/kbfiltr.inx
[the remark this answers is not in the document]
Comment (JakobL-MSFT) [reply input/kbfiltr/sys/kbfiltr.inx:1]: Great, but that is just for testing purposes.  Once testing is complete please remove this change?

File: Build-SampleSet.ps1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (JakobL-MSFT) [reply Build-SampleSet.ps1]: Yes please.

File: .github/scripts/Env-Vars.ps1 [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# source this file to import these environment variables into a session
```
[the remark this answers is not in the document]
Comment (JakobL-MSFT) [reply .github/scripts/Env-Vars.ps1]: Looks great!

File: .github/scripts/Env-Vars.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (JakobL-MSFT) [reply .github/scripts/Env-Vars.ps1]: Thank you. I do like the idea of generating config and prop files, but let's use a different PR for this.

File: .github/workflows/ci-pr.yml (hunk 1 above)
Comment (JakobL-MSFT) [reply .github/workflows/ci-pr.yml:66]: Thank for educating me.

**Review by JakobL-MSFT — CHANGES_REQUESTED** [review]
If you are done with testing I think you should undo any changes to inx files.  Looks like 3 in total!

**Comment by middlemose** [comment]
Issues reverting driver files. Creating a new pull request.

**Pull request closed without merging**